File: src/mutex_bench.zig
const std = @import("std");
const xsync = @import("xsync");
const IoBackend = @import("utils.zig").IoBackend;

// Contended counter: `num_workers` tasks each do `iterations` lock/inc/unlock
// rounds on one shared mutex. Compares std.Io.Mutex (default) with
// xsync.Mutex (--xsync).
const num_workers = 4;
const iterations = 100_000;

fn Bench(comptime Mutex: type) type {
    return struct {
        fn worker(io: std.Io, m: *Mutex, counter: *u64) std.Io.Cancelable!void {
            for (0..iterations) |_| {
                try m.lock(io);
                counter.* += 1;
                m.unlock(io);
            }
        }

        fn run(io: std.Io) !void {
            var m: Mutex = .init;
            var counter: u64 = 0;

            var group: std.Io.Group = .init;
            defer group.cancel(io);
            for (0..num_workers) |_| {
                try group.concurrent(io, worker, .{ io, &m, &counter });
            }
            try group.await(io);

            if (counter != num_workers * iterations) return error.BadCount;
        }
    };
}

pub fn main(init: std.process.Init.Minimal) !void {
    const gpa = std.heap.smp_allocator;

    var backend: IoBackend = .none;
    try backend.init(gpa, init.args);
    defer backend.deinit();

    var use_xsync = false;
    var iter = init.args.iterate();
    while (iter.next()) |arg| {
        if (std.mem.eql(u8, arg, "--xsync")) use_xsync = true;
    }

    const io = backend.io();
    const start: std.Io.Clock.Timestamp = .now(io, .real);

    if (use_xsync) {
        try Bench(xsync.Mutex).run(io);
    } else {
        try Bench(std.Io.Mutex).run(io);
    }

    const duration = start.durationTo(.now(io, .real));
    const ns = duration.raw.toNanoseconds();
    const total: u64 = num_workers * iterations;
    const rate = if (ns > 0) total * 1_000_000_000 / @as(u64, @intCast(ns)) else 0;
    std.log.info("Duration: {f} ({s}, {d} locks, {d} locks/s)", .{ duration.raw, if (use_xsync) "xsync" else "std", total, rate });
}

File: rust/src/bin/mutex_bench.rs
// Counterpart of mutex_bench_native: 4 tasks x 100k lock/inc/unlock on a
// shared std::sync::Mutex (the recommended lock for short critical sections
// in tokio). --async-mutex switches to tokio::sync::Mutex.
use std::sync::{Arc, Mutex};

const NUM_WORKERS: u64 = 4;
const ITERATIONS: u64 = 100_000;

async fn run_std() {
    let counter = Arc::new(Mutex::new(0u64));
    let mut workers = Vec::new();
    for _ in 0..NUM_WORKERS {
        let counter = counter.clone();
        workers.push(tokio::spawn(async move {
            for _ in 0..ITERATIONS {
                *counter.lock().unwrap() += 1;
            }
        }));
    }
    for w in workers {
        w.await.ok();
    }
    assert_eq!(*counter.lock().unwrap(), NUM_WORKERS * ITERATIONS);
}

async fn run_async() {
    let counter = Arc::new(tokio::sync::Mutex::new(0u64));
    let mut workers = Vec::new();
    for _ in 0..NUM_WORKERS {
        let counter = counter.clone();
        workers.push(tokio::spawn(async move {
            for _ in 0..ITERATIONS {
                *counter.lock().await += 1;
            }
        }));
    }
    for w in workers {
        w.await.ok();
    }
    assert_eq!(*counter.lock().await, NUM_WORKERS * ITERATIONS);
}

fn main() {
    let async_mutex = std::env::args().any(|a| a == "--async-mutex");
    let start = std::time::Instant::now();
    let rt = tokio::runtime::Builder::new_multi_thread().enable_all().build().unwrap();
    if async_mutex {
        rt.block_on(run_async());
    } else {
        rt.block_on(run_std());
    }
    let d = start.elapsed();
    println!("Duration: {:.3}ms (tokio {}, {} locks)", d.as_secs_f64() * 1e3, if async_mutex { "async-mutex" } else { "std-mutex" }, NUM_WORKERS * ITERATIONS);
}
